File: B3WM.Shared/Models/VerifierConfig.cs
using B3WM.Shared.Models.Backtest;

namespace B3WM.Shared.Models
{
    public class VerifierConfig
    {
        public string Symbol { get; set; } = "WINFUT";
        public int TimeFrame { get; set; } = 2;
        public StrategyType StrategyName { get; set; } = StrategyType.SmartBreakout;
        public int Quantity { get; set; } = 1;
        public bool IsDayTrade { get; set; } = true;
        public string DayTradeCloseTime { get; set; } = "13:00";
        public int LookbackPeriod { get; set; } = 20;

        // estrutura (arquivos salvos / minDistance)
        public int MinDistance { get; set; } = 0;

        // critérios do SmartBreakout (thresholds de entrada/saída vêm dos
        // filtros de agentes da aba Bubbles, via BubbleThreshold/AgentThresholds)
        public double SmartVolumePct { get; set; } = Defaults.Backtest.SmartVolumePct;
        public double SmartStructureBufferPct { get; set; } = Defaults.Backtest.SmartStructureBufferPct;
        public List<int>? SmartAgents { get; set; }

        // filtro de bubbles (mesmo critério do gráfico: agentes + thresholds)
        public int BubbleThreshold { get; set; } = 0;
        public Dictionary<int, int>? AgentThresholds { get; set; }

        public BacktestConfig ToBacktestConfig() => new()
        {
            Symbol = Symbol,
            TimeFrame = TimeFrame,
            StrategyName = StrategyName,
            Quantity = Quantity,
            CommissionPerSide = 0,
            IsDayTrade = IsDayTrade,
            DayTradeCloseTime = DayTradeCloseTime,
            LookbackPeriod = LookbackPeriod,
            MinDistance = MinDistance,
            SmartVolumePct = SmartVolumePct,
            SmartStructureBufferPct = SmartStructureBufferPct,
            SmartAgents = SmartAgents,
            BubbleThreshold = BubbleThreshold,
            AgentThresholds = AgentThresholds
        };
    }
}

File: B3WM.Flutter/lib/models/verifier_config.dart
class VerifierConfig {
  static const List<String> strategyTypes = ['SmartBreakout', 'Breakout'];

  String symbol;
  int timeFrame;
  String strategyName;
  int quantity;
  bool isDayTrade;
  String dayTradeCloseTime;
  int lookbackPeriod;
  int minDistance;
  double smartVolumePct;
  double smartStructureBufferPct;
  List<int>? smartAgents;
  int bubbleThreshold;
  Map<int, int>? agentThresholds;

  VerifierConfig({
    this.symbol = 'WINFUT',
    this.timeFrame = 2,
    this.strategyName = 'SmartBreakout',
    this.quantity = 1,
    this.isDayTrade = true,
    this.dayTradeCloseTime = '13:00',
    this.lookbackPeriod = 20,
    this.minDistance = 0,
    this.smartVolumePct = 0.3,
    this.smartStructureBufferPct = 0.1,
    this.smartAgents,
    this.bubbleThreshold = 0,
    this.agentThresholds,
  });

  factory VerifierConfig.fromJson(Map<String, dynamic> json) {
    return VerifierConfig(
      symbol: json['symbol'] as String? ?? 'WINFUT',
      timeFrame: (json['timeFrame'] as num?)?.toInt() ?? 2,
      strategyName: json['strategyName'] as String? ?? 'SmartBreakout',
      quantity: (json['quantity'] as num?)?.toInt() ?? 1,
      isDayTrade: json['isDayTrade'] as bool? ?? true,
      dayTradeCloseTime: json['dayTradeCloseTime'] as String? ?? '13:00',
      lookbackPeriod: (json['lookbackPeriod'] as num?)?.toInt() ?? 20,
      minDistance: (json['minDistance'] as num?)?.toInt() ?? 0,
      smartVolumePct: (json['smartVolumePct'] as num?)?.toDouble() ?? 0.3,
      smartStructureBufferPct:
          (json['smartStructureBufferPct'] as num?)?.toDouble() ?? 0.1,
      smartAgents: json['smartAgents'] != null
          ? (json['smartAgents'] as List).map((e) => (e as num).toInt()).toList()
          : null,
      bubbleThreshold: (json['bubbleThreshold'] as num?)?.toInt() ?? 0,
      agentThresholds: json['agentThresholds'] != null
          ? (json['agentThresholds'] as Map<String, dynamic>)
              .map((k, v) => MapEntry(int.tryParse(k) ?? 0, (v as num).toInt()))
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'symbol': symbol,
        'timeFrame': timeFrame,
        'strategyName': strategyName,
        'quantity': quantity,
        'isDayTrade': isDayTrade,
        'dayTradeCloseTime': dayTradeCloseTime,
        'lookbackPeriod': lookbackPeriod,
        'minDistance': minDistance,
        'smartVolumePct': smartVolumePct,
        'smartStructureBufferPct': smartStructureBufferPct,
        'smartAgents': smartAgents,
        'bubbleThreshold': bubbleThreshold,
        'agentThresholds':
            agentThresholds?.map((k, v) => MapEntry(k.toString(), v)),
      };

  VerifierConfig copyWith({
    String? symbol,
    int? timeFrame,
    String? strategyName,
    int? quantity,
    bool? isDayTrade,
    String? dayTradeCloseTime,
    int? lookbackPeriod,
    int? minDistance,
    double? smartVolumePct,
    double? smartStructureBufferPct,
    List<int>? smartAgents,
    int? bubbleThreshold,
    Map<int, int>? agentThresholds,
  }) {
    return VerifierConfig(
      symbol: symbol ?? this.symbol,
      timeFrame: timeFrame ?? this.timeFrame,
      strategyName: strategyName ?? this.strategyName,
      quantity: quantity ?? this.quantity,
      isDayTrade: isDayTrade ?? this.isDayTrade,
      dayTradeCloseTime: dayTradeCloseTime ?? this.dayTradeCloseTime,
      lookbackPeriod: lookbackPeriod ?? this.lookbackPeriod,
      minDistance: minDistance ?? this.minDistance,
      smartVolumePct: smartVolumePct ?? this.smartVolumePct,
      smartStructureBufferPct:
          smartStructureBufferPct ?? this.smartStructureBufferPct,
      smartAgents: smartAgents ?? this.smartAgents,
      bubbleThreshold: bubbleThreshold ?? this.bubbleThreshold,
      agentThresholds: agentThresholds ?? this.agentThresholds,
    );
  }
}
